march_hardware/src/IMotionCube.cpp

```diff
@@ -84,9 +84,19 @@ void IMotionCube::mapMosiPDOs()
 }
 
 // Set configuration parameters to the IMC
-void IMotionCube::writeInitialSettings(uint8_t cycle_time)
+bool IMotionCube::writeInitialSettings(uint8_t cycle_time)
 {
-  ROS_DEBUG("IMotionCube::writeInitialSettings");
+  int checksum_verified = verifySetup();
+  ROS_INFO("The .sw file for slave %d is %s", this->slaveIndex,
+           checksum_verified ? "equal to the setup of the drive." :
+                               "not equal to the setup of the drive, downloading is necessary");
+  if (!checksum_verified)
+  {
+    downloadSetupToDrive();
+    checksum_verified = verifySetup();
+    ROS_INFO(checksum_verified ? "writing of the setup data has succeeded" : "writing of the setup data has failed");
```

Comment:
Isn't this more severe, may be even fatal. Why know there is a wrong iMC project and we are not able to recover. Should we really continue?

Reply:
I want the system to be able to recover from a potential msitake. Making it fatal would kill the node and I want to have a robust way of downloading these files. We have encountered many experts who suggest to just try again and see if the same thing happens before you jump to conclusions after a single failure.

Reply:
From what I saw on Wednesday, this does not really fail (often). I think you can just make this a fatal and remove the loop at in MarchRobot.cpp (so only try once). If it fatals very rarely, we can simply launch again. If this fatals too often we can always implement the loop again.

Reply:
fair point. I'll implement these changes.